### PR TITLE
Add Forchheimer IPR Case Description 

### PR DESCRIPTION
**Pull request by caio-ramos**
ASIM-3663

Add a `CaseDescription` of the Forchheimer IPR model.

### CODE REVIEW

**Comment by nicoddemus** [comment]
We will need to wait for you to make a new barril release before moving forward with this then. :+1: 

- [ ] Wait for new `barril` release

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (ggrbill) [reply src/alfasim_sdk/_internal/alfacase/case_description.py:1195]: Can you name those options?

I suggest adding a new enum (`ForchheimerCoefficientsOption`) that will have the options `From User` and `Calculated from experimental data` and add a prefix `exp_data_` to the variables that will be used to calculate the coefficients.

What do you think?
I am not sure if my suggestion makes sense here. If not, please let me know.

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (ggrbill) [inline src/alfasim_sdk/_internal/alfacase/case_description.py:1163]: why no `gas_viscosity`?

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (ggrbill) [inline src/alfasim_sdk/_internal/alfacase/case_description.py:1164]:
```suggestion
    gas_z_factor = attrib_scalar(default=Scalar("dimensionless", 0.0, "-"))
```

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (caio-ramos) [reply src/alfasim_sdk/_internal/alfacase/case_description.py]: Your suggestion was very helpful. I did something very similar, please check it out and let me know if you agree.

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (caio-ramos) [reply src/alfasim_sdk/_internal/alfacase/case_description.py]: No problem, done!

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (caio-ramos) [reply src/alfasim_sdk/_internal/alfacase/case_description.py]: Thanks, done!

**Comment by caio-ramos** [comment]
Unfortunately, I couldn't run `inv docs` locally, which might give some clue as to what's going on with `docs`. Can anyone help me with this? What am I missing here?
```
InputError: [Errno 2] No such file or directory: 'alfacase_definitions/list_of_unit_for_forchheimer_linear_productivity_index.txt'.
InputError: [Errno 2] No such file or directory: 'alfacase_definitions/list_of_unit_for_forchheimer_quadratic_productivity_index.txt'.
InputError: [Errno 2] No such file or directory: 'alfacase_definitions/list_of_unit_for_permeability_rock.txt'.
InputError: [Errno 2] No such file or directory: 'alfacase_definitions/list_of_unit_for_nonDarcy_flow_coefficient.txt'.
```
Do I need to create these files manually or is there a command to generate them?

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (ggrbill) [inline src/alfasim_sdk/_internal/alfacase/case_description.py:1163]: Nice!

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (ggrbill) [inline src/alfasim_sdk/_internal/alfacase/case_description.py:1776]: Something happened here. The `η` symbol was removed. Please, fix this.

File: src/alfasim_sdk/_internal/constants.py
Comment (ggrbill) [inline src/alfasim_sdk/_internal/constants.py:535]: 👍🏼

File: tests/common_testing/alfasim_sdk_common_testing/filled_case_descriptions.py
Comment (ggrbill) [inline tests/common_testing/alfasim_sdk_common_testing/filled_case_descriptions.py:362]: I suggest adding two usages of Forchheimer IPR:
```python
FORCHHEIMER_IPR_DESCRIPTION = case_description.ForchheimerIPRDescription(
    well_index_phase=constants.WellIndexPhaseType.Oil,
    min_pressure_difference=Scalar(0.0, "bar"),
    calculate_coeff_option=constants.ForchheimerCoefficientsOption.ReservoirParameters,
    gas_viscosity=Scalar("dynamic viscosity", 0.0, "Pa.s"),
    gas_z_factor=Scalar("dimensionless", 0.0, "-"),
    reservoir_permeability=Scalar("permeability rock", 0.0, "m2"),
    drainage_radius=Scalar("length", 0.0, "m"),
    well_radius=Scalar("length", 0.0, "m"),
    well_skin_factor=Scalar("dimensionless", 0.0, "-"),
    non_darcy_parameter=Scalar("nonDarcy flow coefficient", 0.0, "Pa.s/m6"),
)
FORCHHEIMER_IPR_COEFF_DESCRIPTION = case_description.ForchheimerIPRDescription(
    well_index_phase=constants.WellIndexPhaseType.Oil,
    min_pressure_difference=Scalar(0.0, "bar"),
    calculate_coeff_option=constants.ForchheimerCoefficientsOption.FlowCoefficients,
    B_coeff=Scalar(1.0, "Pa2.s/scm"),
    C_coeff=Scalar(1.0, "Pa2.s2/scm2"),
)
```
With that, we would have both ways of Forchheimer IPR configuration. What do you think?

**Review by ggrbill — APPROVED** [review]
Good Job!

Please, before merging fix the `η` symbol changed.

File: tests/common_testing/alfasim_sdk_common_testing/filled_case_descriptions.py
Comment (caio-ramos) [reply tests/common_testing/alfasim_sdk_common_testing/filled_case_descriptions.py:362]: This is a good idea, but the case description `FORCHHEIMER_IPR_COEFF_DESCRIPTION` will not be tested. If you look at [test_alfacase_to_case.py](https://github.com/ESSS/alfasim-sdk/pull/313/files#diff-039440478d6ecc71845eb828ddc2e38fc2284e7c98d04b8a9162e0128fb7cdc9bR277-R281) you will find a dictionary where the key is `ForchheimerIPRDescription`, which must match the name of the class description. If I want to test `FORCHHEIMER_IPR_DESCRIPTION` and `FORCHHEIMER_IPR_COEFF_DESCRIPTION` a class is needed for each option, which goes against the initial idea of ​​having a single class with all the data.

Comment (ggrbill) [reply tests/common_testing/alfasim_sdk_common_testing/filled_case_descriptions.py:362]: Hmm, I understand. Then, I suggest adding only the model data that calc would use (One of two descriptions that I suggested before).  Is it possible? Or the test would check all possible description parameters?

Comment (caio-ramos) [reply tests/common_testing/alfasim_sdk_common_testing/filled_case_descriptions.py:362]: It is done.  Since we are assuming the `ReservoirParameters` as default option, I removed the `B_coeff` and `C_coeff`. Removing these arguments makes them take on their own default value, so I think we're good.

**Comment by caio-ramos** [comment]
> > Do I need to create these files manually or is there a command to generate them?
> 
> I am not sure, but since it is for documentation I would say you have to add it manually.

`inv cog` should handle it, adding these new categories to this file everything works fine src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py

File: docs/source/alfacase_definitions/list_of_unit_for_nonDarcy_flow_coefficient.txt
Comment (prusse-martin) [inline docs/source/alfacase_definitions/list_of_unit_for_nonDarcy_flow_coefficient.txt]: Why there are 2 identical entries?

**Review by prusse-martin — CHANGES_REQUESTED** [review]

File: src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py
Comment (prusse-martin) [inline src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py:93]:
```suggestion
        *sorted(body, key=lambda x: (x.casefold(), x)),
```

**Review by prusse-martin — CHANGES_REQUESTED** [review]

File: src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py
Comment (caio-ramos) [reply src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py]: Thanks for the careful review, it's all done!

File: src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py
Comment (caio-ramos) [reply src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py:89]: Yes, I haven't found a better solution than your recommendation. But is all right now.

File: src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py]: `*sorted(body, key=lambda x: (x.casefold(), x.swapcase())),` will put the ones with lower case first, but I guess that is marginal change =)

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (prusse-martin) [reply src/alfasim_sdk/_internal/alfacase/case_description.py:1163]: Just
```
    calculate_coeff_option = attrib_enum(
        default=constants.ForchheimerCoefficientsOption.ReservoirParameters,
    )
```
will have the same effect, when givin `default` the `type_` is ignored. Maybe we should at least check if when using both the match.

**Review by nicoddemus — APPROVED** [review]

File: src/alfasim_sdk/_internal/alfacase/case_description.py
Comment (caio-ramos) [reply src/alfasim_sdk/_internal/alfacase/case_description.py]: Thanks. Now I'm follow your suggestion

File: src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py
Comment (caio-ramos) [reply src/alfasim_sdk/_internal/alfacase/generate_case_description_docstring.py]: I'm with you, having lowercase letters first is better presentable. It's done.

**Review by prusse-martin — APPROVED** [review]

**Pull request merged**
